cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/AuthorizationModule.java

```diff
@@ -0,0 +1,130 @@
+    bind(Admin.class).toProvider(AdminProvider.class);
+    bind(Transactional.class).toProvider(TransactionalProvider.class).in(Scopes.SINGLETON);
+
+    bind(AuthorizerInstantiatorService.class).in(Scopes.SINGLETON);
+    expose(AuthorizerInstantiatorService.class);
+  }
+
+  private static final class DynamicDatasetCacheProvider implements Provider<DynamicDatasetCache> {
+
+    private final DatasetFramework dsFramework;
+    private final TransactionSystemClient txClient;
+    private final MetricsCollectionService metricsCollectionService;
+
+    @Inject
+    private DynamicDatasetCacheProvider(DatasetFramework dsFramework, TransactionSystemClient txClient,
+                                        MetricsCollectionService metricsCollectionService) {
+      this.dsFramework = dsFramework;
+      this.txClient = txClient;
+      this.metricsCollectionService = metricsCollectionService;
+    }
+
+    @Override
+    public DynamicDatasetCache get() {
+      SystemDatasetInstantiator dsInstantiator = new SystemDatasetInstantiator(dsFramework, null, null);
+      return new MultiThreadDatasetCache(
+        dsInstantiator, txClient, NamespaceId.SYSTEM, ImmutableMap.<String, String>of(),
+        metricsCollectionService.getContext(ImmutableMap.<String, String>of()),
+        ImmutableMap.<String, Map<String, String>>of()
+      );
+    }
+  }
+
+  private static final class AdminProvider implements Provider<Admin> {
+    private final DatasetFramework dsFramework;
+
+    @Inject
+    private AdminProvider(DatasetFramework dsFramework) {
+      this.dsFramework = dsFramework;
+    }
+
+    @Override
+    public Admin get() {
+      return new DefaultAdmin(dsFramework, NamespaceId.SYSTEM);
+    }
+  }
+
+  private static final class TransactionalProvider implements Provider<Transactional> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(TransactionalProvider.class);
+
+    private final DynamicDatasetCache datasetCache;
+
+    @Inject
+    private TransactionalProvider(DynamicDatasetCache datasetCache) {
+      this.datasetCache = datasetCache;
+    }
+
+    @Override
+    public Transactional get() {
+      return new Transactional() {
+        @Override
+        public void execute(TxRunnable runnable) throws TransactionFailureException {
+          TransactionContext transactionContext = datasetCache.newTransactionContext();
+          transactionContext.start();
+          try {
+            runnable.run(datasetCache);
+          } catch (TransactionFailureException e) {
+            LOG.error("Exception while executing runnable inside a transaction. Aborting transaction.", e);
+            transactionContext.abort(e);
+          } catch (Throwable t) {
+            LOG.error("Exception while executing runnable inside a transaction. Aborting transaction.", t);
+            transactionContext.abort(new TransactionFailureException("Exception raised from TxRunnable.run()", t));
+          }
+          transactionContext.finish();
```

Comment:
For readability, it would be better to move `transactionContext.finish()` after `runnable.run()`